### PR TITLE
Make loot accurate to the loot drop rates during live. 

### PR DESCRIPTION
**Pull request by Jettford**
This PR is mostly made of repaired code from @averysumner that was never finished and has now been reimplemented.

I have no idea of knowing if these drops are accurate however Krysto (a prominent community member, who knows much more about loot than I) had spent a good amount of time with this build testing the drop rates.

### CODE REVIEW

**Comment by Jettford** [comment]
Tested on Windows 11 (Clang) and Ubuntu 20 (GCC)

**Comment by averysumner** [comment]
I’d like @NealSpellman’s input on these drops before it’s merged.

**Review by DarwinAnim8or — APPROVED** [review]
The code looks ok for me.  Didn't try to run anything wrt look accuracy though. 
@averysumner should be the one to have final say here

**Comment by NealSpellman** [comment]
I don't have the time this week to test this, but I'd like someone or Krysto to chime in with data on and confirmations around all loot cases.
- Item packages
    - Packages with predetermined contents
    - Packages with random contents
- Boss loot
- Activity loot
- Team loot
- Wishing wells
- Mission loot (running a mission that requires drops from an enemy to confirm this didn't break)
...etc

This is a really major thing that we don't want broken on main, so someone will need to run this and test a _very wide_ variety of loot sources so we don't break loot for everyone. Preferably also running a simulator on many of the loot pools, such as bosses, to confirm this logic is working right and the odds are roughly the same but there still needs to be a lot of in-game testing in addition.

**Comment by Jettford** [comment]
Alright I will get Krysto to chime in

**Comment by Jettford** [comment]
Also team loot is completely broken but it is on main right now so.... nothing hurt

**Comment by Jettford** [comment]
Alright I have a list of things tested by Krysto accompanied by a statement from him

- Surprise packs
- Consistent packs
- Boss loot
- Mission-specific loot
- Wishing wells

> If there's anything wrong with them, it would require extensive and exhaustive testing beyond the abilities of a single person
at a glance I have found no immediate or glaring issues

**Comment by NealSpellman** [comment]
Still will need more testing than just one person, but it's good the initial results from Krysto are positive. This needs further testing from others and actual drop simulation work on your part to prove this is (roughly) matching what can be calculated from the database.

**Comment by Verathian2** [comment]
Hey all,
I have found on my own server that the drops, especially from gorillas and admirals in GF have felt lower than the original game. Farmed admirals for a couple hours and only got a couple of shirts, pants, etc. (the kind of fillers you would get from dragons in FV). Not that you shouldn't get those things, but that's all I've been seeing.

**Comment by averysumner** [comment]
> Hey all,
> I have found on my own server that the drops, especially from gorillas and admirals in GF have felt lower than the original game. Farmed admirals for a couple hours and only got a couple of shirts, pants, etc. (the kind of fillers you would get from dragons in FV). Not that you shouldn't get those things, but that's all I've been seeing.

It’s been a decade, are you basing this on feeling or video evidence?

**Comment by averysumner** [comment]
> Still will need more testing than just one person, but it's good the initial results from Krysto are positive. This needs further testing from others and actual drop simulation work on your part to prove this is (roughly) matching what can be calculated from the database.

@Jettford maybe you could add a drop simulator command to this?

**Comment by Jettford** [comment]
> @Jettford maybe you could add a drop simulator command to this?

Yeah I can add a command to simulate loot drops. Will add that now.

**Comment by Jettford** [comment]
Implemented a `/rollloot` command to the server (do not run this on a live server, it will basically kill it) the format is as follows

`/rollloot <LootMatrixIndex> <Item ID> <times>`

e.g. 
`/rollloot 218 7570 1000`: Runs 1000 times gathering amount of runs it takes for a Stromling Ape (LMI: 218) to drop a parrot (item id: 7570)

It will provide an output as follows:
```
Ran loot drops looking for 7570, 1000 times. It ran 17467393 times. Averaging out at 17367.392578
```

**Comment by averysumner** [comment]
> Implemented a `/rollloot` command to the server (do not run this on a live server, it will basically kill it) the format is as follows
> 
> `/rollloot <LootMatrixIndex> <Item ID> <times>`
> 
> e.g. `/rollloot 218 7570 1000`: Runs 1000 times gathering amount of runs it takes for a Stromling Ape (LMI: 218) to drop a parrot (item id: 7570)
> 
> It will provide an output as follows:
> 
> ```
> Ran loot drops looking for 7570, 1000 times. It ran 17467393 times. Averaging out at 17367.392578
> ```

At least for the red parrot, that's good convergence on the expected value of 17500 (if my math is correct). I would like some testing with some other objects too (dragons, stromlings, different tiers of enemies, etc).

**Review by MickVermeulen — APPROVED** [review]
Agree with Darwin, although I do also agree on this needing more testing.

File: dGame/dUtilities/Loot.cpp
Comment (Wincent01) [inline dGame/dUtilities/Loot.cpp:95]: This will result in a lot of double storage. Consider using a ptr to the already cached table entry.

**Review by Wincent01 — CHANGES_REQUESTED** [review]
Believe there is still some loot logic to be ironed out. You are able to get large amounts of duplicates from boss enemies like Butterscorch. Got four "Iron Barbute" once, and four "Denim Shirt" along with three "Block Shirt" another time.

**Comment by NealSpellman** [comment]
> Believe there is still some loot logic to be ironed out. You are able to get large amounts of duplicates from boss enemies like Butterscorch. Got four "Iron Barbute" once, and four "Denim Shirt" along with three "Block Shirt" another time.

Yeah, for this stuff I'd suggest watching videos from Live about players smashing dragons and the loot they pick up. This seems broken if they're getting _this many_ duplicates.

**Comment by Jettford** [comment]
> Yeah, for this stuff I'd suggest Jett watches videos from Live about players smashing dragons and the loot they pick up. This seems broken if they're getting this many duplicates.

I don't think this is the case, I think it is mainly down to you are smashing dragons solo

**Comment by Jettford** [comment]
3-4 items seems right considering that is how the loot was kind of supposed to be divided up between teams.

**Comment by NealSpellman** [comment]
> > Yeah, for this stuff I'd suggest Jett watches videos from Live about players smashing dragons and the loot they pick up. This seems broken if they're getting this many duplicates.
> 
> I don't think this is the case, I think it is mainly down to you are smashing dragons solo

For the LMI and activity related to a solo smash, you shouldn't be getting nearly that. Dragons do not drop the same amount of items for solo smashes as they would for higher groups. If this is indeed broken on main right now, then I need data on each of the LMIs that the different varying player numbers have to see if this works as expected.

**Comment by Jettford** [comment]
So are you saying that the loot that should be dropped should be 1/4 of what it is right now?

**Comment by NealSpellman** [comment]
> So are you saying that the loot that should be dropped should be 1/4 of what it is right now?

I'm saying there's 4 different LMIs for a regular dragon on Crux.

`{ "LootMatrixIndex" : 512, "Description" : "Crux Prime: Maelstrom Dragon Invader"},
{ "LootMatrixIndex" : 677, "Description" : "Crux Prime: Maelstrom Dragon Invader, 2 Players"},
{ "LootMatrixIndex" : 678, "Description" : "Crux Prime: Maelstrom Dragon Invader, 3 Players"},
{ "LootMatrixIndex" : 679, "Description" : "Crux Prime: Maelstrom Dragon Invader, 4 Players"},
`

I'm not sure what main branch is doing for it right now / if we take into account player count for regular dragon rewards, but the current results from this ticket don't sound like they're 1 player rewards. If this is broken on main, I'd settle for simulation data confirming all of these are working as intended if you don't intend to fix existing loot bugs.

**Comment by Jettford** [comment]
ah this is interesting, I will take a look at the object data to work out how I can get these values on the server and fix it, but this issue is present on main right now

**Comment by Jettford** [comment]
Alright I have gone through it, it is dropping 1 player loot and it appears to be correct

**Comment by Jettford** [comment]
The only issue is the duplicates so I think your duplicate protection idea is worth looking into

File: dGame/dUtilities/Loot.cpp
Comment (Jettford) [reply dGame/dUtilities/Loot.cpp:95]: I think the speed of not checking through them on add outweighs the few kb of lost memory

**Comment by averysumner** [comment]
This is significantly better than what we have now, I'm merging this.

**Pull request merged**
